stacks-signer/src/crypto/frost.rs

```diff
@@ -0,0 +1,888 @@
+
+    fn start_private_shares(&mut self) -> Result<Message, Error> {
+        info!(
+            "DKG Round #{}: Starting Private Share Distribution",
+            self.current_dkg_id
+        );
+        let dkg_begin = DkgBegin {
+            dkg_id: self.current_dkg_id,
+        };
+        let dkg_private_begin_msg = Message {
+            sig: dkg_begin.sign(&self.message_private_key).expect(""),
+            msg: MessageTypes::DkgPrivateBegin(dkg_begin),
+        };
+        self.move_to(State::DkgEndGather)?;
+        Ok(dkg_private_begin_msg)
+    }
+
+    fn gather_public_shares(&mut self, message: &Message) -> Result<(), Error> {
+        match &message.msg {
+            MessageTypes::DkgPublicEnd(dkg_end) => {
+                if dkg_end.dkg_id != self.current_dkg_id {
+                    return Err(Error::BadDkgId(dkg_end.dkg_id, self.current_dkg_id));
+                }
+                self.ids_to_await.remove(&dkg_end.signer_id);
+                info!(
+                    "DKG_Public_End round #{} from signer #{}. Waiting on {:?}",
+                    dkg_end.dkg_id, dkg_end.signer_id, self.ids_to_await
+                );
+            }
+            MessageTypes::DkgPublicShare(dkg_public_share) => {
+                if dkg_public_share.dkg_id != self.current_dkg_id {
+                    return Err(Error::BadDkgId(
+                        dkg_public_share.dkg_id,
+                        self.current_dkg_id,
+                    ));
+                }
+                if dkg_public_share.dkg_public_id != self.current_dkg_public_id {
+                    return Err(Error::BadDkgPublicId(
+                        dkg_public_share.dkg_public_id,
+                        self.current_dkg_public_id,
+                    ));
+                }
+                self.dkg_public_shares
+                    .insert(dkg_public_share.party_id, dkg_public_share.clone());
+
+                info!(
+                    "DKG round #{} DKG public round #{} DkgPublicShare from party #{}",
+                    dkg_public_share.dkg_id,
+                    dkg_public_share.dkg_public_id,
+                    dkg_public_share.party_id
+                );
+            }
+            _ => {}
+        }
+        if self.ids_to_await.is_empty() {
+            // Calculate the aggregate public key
+            let key = self
+                .dkg_public_shares
+                .iter()
+                .fold(Point::default(), |s, (_, dps)| s + dps.public_share.A[0]);
+            // check to see if aggregate public key has even y
+            if key.has_even_y() {
+                info!("Aggregate public key has even y coord!");
+                info!("Aggregate public key: {}", key);
+                self.aggregate_public_key = key;
+                self.move_to(State::DkgPrivateDistribute)?;
+            } else {
+                warn!("DKG Round #{} Failed: Aggregate public key does not have even y coord, re-running dkg.", self.current_dkg_id);
```

Comment:
I remember a while back we were discussing this possible outcome.  I don't remember this point, though: can this happen by accident, or does this condition only arise from malicious behavior on the coordinator's part?

Either way the signers need to re-run DKG, but if this only happens due to malicious behavior, then should the logic here identify the malicious actor and exclude them from the next DKG round?

Also, regarding the comment below -- does _not_ incrementing the `dkg_public_id` (or having some way of tracking the distribute/gather instance) mean that some signers' gathered public shares could be from a "stale" distribute/gather cycle?  If so, is that a problem?

Reply:
Ah nm, I see this was discussed in another comment thread :pray: